spoc/snipping/snipping_strategies.py

```diff
@@ -0,0 +1,237 @@
+"""Snipping strategies for Snipper that implement specific snipping functionality"""
+from abc import ABC, abstractmethod
+from functools import partial
+from enum import Enum
+from typing import Union
+from multiprocess.pool import ThreadPool
+import pandas as pd
+import numpy as np
+from sparse import COO
+import duckdb
+import bioframe
+from spoc.pixels import PersistedPixels
+
+
+class SnippingValues(Enum):
+    """Which values the snippet should consist of."""
+
+    ICCF = 0
+    OBSEXP = 1
+
+
+class SnippingStrategy(ABC):
+    """Defines interface for snipping strategy"""
+
+    def __init__(
+        self,
+        bin_size: int,
+        half_window_size: int,
+        snipping_value: SnippingValues,
+        **kwargs,
+    ):
+        """Defines the values that should be snipped"""
+        self._snipping_value = snipping_value
+        self._half_window_size = half_window_size
+        if snipping_value == SnippingValues.OBSEXP:
+            self._n_random_regions: int = kwargs.get("n_random_regions", 5000)
+            self._genome: str = kwargs.get("genome", "hg19")
+        self._bin_size = bin_size
+
+    @staticmethod
+    def _get_random_coordinates(n_coordinates: int, length: int, genome: str):
```

Comment:
This is not a bad background model, but not the best. We'd better use shuffling procedures from GenometriCorr. Specifically, they have a procedure, that shuffles genomic intervals while preserving their lengths and gaps between them. I have an implementation of this procedure in python for a single chromosome: [https://gist.github.com/dmitrymyl/64c88d7050faf76a62a6744e552a47f8](url). I refactored GenometriCorr procedure to accommodate for discovered corner cases (specifically for deeply nested intervals), but it still needs testing.

This will also change the way user supply data: instead of points user should give genomic intervals in a form of a dataframe, whily SnippingStrategy should extract boundaries (in case of TADs) or centers (in case of narrowPeaks).

There should also be a SnippingStrategy for single points, but _get_random_coordinates should produce the same number of random points and ideally preserve distances between them.

Reply:
That is an interesting thought! I think it is a question of what we want the expected model to represent. This implementation represents a model that does not assume any structure in the supplied regions. That means that the expected model captures the dependence of contact frequency on genomic distance configuration and any difference to it highlights structure in the supplied regions. This is in line with what conventional Hi-C expectation calculations do, so it might be more familiar to people interpreting the data. If we incorporate knowledge of the structure of the regions in the expected calculations, I think the resulting data becomes harder to interpret. I get why it is needed for comparing intervals (as in GenometriCorr), but I would be cautious to adapt this here. Let's discuss this in person :)